django/db/models/sql/compiler.py

```diff
@@ -1331,24 +1330,34 @@ def as_sql(self):
             ]
 
     def execute_sql(self, return_id=False):
-        assert not (
-            return_id and len(self.query.objs) != 1 and
-            not self.connection.features.can_return_rows_from_bulk_insert
+        return_id = (
+            not return_id and
+            len(self.query.objs) == 1 or
+            self.connection.features.can_return_rows_from_bulk_insert
         )
         self.return_id = return_id
         with self.connection.cursor() as cursor:
             for sql, params in self.as_sql():
                 cursor.execute(sql, params)
-            if not return_id:
-                return
+            if not (return_id and self.query.get_meta().db_returning_fields):
+                return []
             if self.connection.features.can_return_rows_from_bulk_insert and len(self.query.objs) > 1:
-                return self.connection.ops.fetch_returned_insert_ids(cursor)
+                return self.connection.ops.fetch_returned_insert_rows(cursor)
             if self.connection.features.can_return_columns_from_insert:
                 assert len(self.query.objs) == 1
-                return self.connection.ops.fetch_returned_insert_id(cursor)
-            return self.connection.ops.last_insert_id(
-                cursor, self.query.get_meta().db_table, self.query.get_meta().pk.column
-            )
+                return self.connection.ops.fetch_returned_insert_columns(cursor)
+            meta = self.query.get_meta()
+            if meta.pk.get_internal_type() in ['AutoField', 'BigAutoField', 'SmallAutoField']:
```

Comment:
Prefer a set (or tuple) to list for containment check.

Reply:
This feels like a kludge. Is it dependant on something like the changes in #3776?

Reply:
Good catch, a set is particularly fast if the element is not contained, which is the larger case.

And yes, it's a kludge, but no one has come up with a better solution yet. It would need some major refactoring of the database backends. Which I am happy to do, but not as part of this PR.